packages/build/src/plugins/compatibility.ts

```diff
@@ -102,8 +102,19 @@ const getCompatibleEntry = async function ({
     )
   })
 
+  if (compatibleEntry) {
+    return compatibleEntry
+  }
+
+  if (pinnedVersion) {
+    return { version: pinnedVersion, conditions: [] }
+  }
+
   return (
-    compatibleEntry ||
-    (pinnedVersion ? { version: pinnedVersion, conditions: [] } : { version: versions[0].version, conditions: [] })
+    (await pLocate(versions, async ({ conditions }) => {
+      return await pEvery(conditions, async ({ type, condition }) =>
+        CONDITIONS[type].test(condition as any, { nodeVersion, packageJson, packagePath, buildDir }),
+      )
+    })) ?? { version: versions[0].version, conditions: [] }
```

Comment:
the fix here is to try to find compatible entry (for condititions / constraints) in case `pinnedVersion` is not set instead of picking first one (which might not satisfy conditions)